### PR TITLE
Math-Magicians: Add Events

### PR DESCRIPTION
**Pull request by ZikaZaki**
# Math-Magicians: Add Events

## In this pull request I did the followings:

- [x] Created a **directory** called `context` for managing all the app **contexts**.
- [x] Inside **context** directory, created a `CalcContext.js` for creating the calculator context.
- [x] Added the [react textfit](https://www.npmjs.com/package/react-textfit) library to improve the UX of the calculator `screen`.
- [x] Added a `Button.js` file to the **components** directory for creating the button component.
- [x] Added `getClassName()` method for fetching the **CSS ClassName** of the button.
- [x] Created a `Logic` directory & copied the [calculate.js](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/code-samples/calculate.js) and [operate.js](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/code-samples/operate.js) files into it.
- [x] Added `handleClick()` event for handling all click events.
- [x] Used `ES6` for all the created components (All components are **class-based**).
- [x] Assembled all the **Calculator** components into the main `Calculator.js` compoent, then imported the main `Calculator.js`  component into the **`App.js`** component.
- [x] Tested the **Calculator** in the browser and it rendered without problems.
- [x] Tested all the **Calculator functionality** and all of them worked as intended.
- [x] Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
- [x] Made sure that the project is perfectly working as intended.
- [x] Documented the project professionally.

### CODE REVIEW

File: src/components/CalcScreen.js [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-    const { value } = this.props;
+    const { calc } = this.context;
     return (
-      <div className="calcScreen">
-        {value}
-      </div>
+      <Textfit className="calcScreen" max={50} mode="single">
+        {calc.num ? calc.num : calc.res}
+      </Textfit>
     );
```
Comment (BenMukebo) [inline src/components/CalcScreen.js]: - [x] Great job so far 👏🏻 However, according to the project requirements. To make work easy and concentrate on how to work with React events, you are advised to use the already developed `math logic` needed to make work the Calculator component. Let's make sure that:
*  You hava an `operate.js` file copied from [this one](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/code-samples/operate.js).
* You also have a `calculate.js` file copied from [this one](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/code-samples/calculate.js)
- You need to create a `logic` forder directory for the  `operate.js` and the `calculate.js` files.

File: src/components/Calculator.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ import CalcWrapper from './CalcWrapper';
 import CalcScreen from './CalcScreen';
 import ButtonsPad from './ButtonsPad';
 import Button from './Button';
+import CalcProvider from '../context/CalcContext';
```
Comment (BenMukebo) [inline src/components/Calculator.js:6]: - [x] As said above 🙃  The `Calculator` component should import the `calculate.js` file from the logic directory:

```javascript
// ./src/components/Calculator.js

import React, { Component } from 'react';
import calculate from ../logic/calculator;
```

   * `operate.js` is already being used by `calculate.js`, so there's no need to import it

File: src/components/Calculator.js [hunk 3]
```diff
@@ -18,14 +19,16 @@ class Calculator extends React.Component {
...
-        </ButtonsPad>
-      </CalcWrapper>
+      <CalcProvider>
+        <CalcWrapper>
+          <CalcScreen value="0" />
+          <ButtonsPad>
+            {this.btnValues.flat().map((row, index) => (
+              <Button key={index} value={row} />
+            ))}
+          </ButtonsPad>
```
Comment (BenMukebo) [inline src/components/Calculator.js:29]: - [x] You need to use `state` and `props` and make that props get passed to child components of the `Calculate.js` component.

- I know you can do it 💪🏻 Great job 👍🏻

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @ZikaZaki,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻
- No linter errors. ✔️ 
- Gitflow was correctly used.  ✔️   
- Good commit messages ✔️ 
- Descriptive README file.  ✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/CalcScreen.js (hunk 1 above)
Comment (ZikaZaki) [reply src/components/CalcScreen.js]: Thanks Mr. @BenMukebo  for the great feedback. I have already implemented the requested changes.

File: src/components/Calculator.js (hunk 2 above)
Comment (ZikaZaki) [reply src/components/Calculator.js:6]: Thanks Mr. @BenMukebo  for the great feedback. I have already implemented the requested changes.

File: src/components/Calculator.js (hunk 3 above)
Comment (ZikaZaki) [reply src/components/Calculator.js:29]: Thanks Mr. @BenMukebo  for the great feedback. I have already implemented the requested changes.

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @ZikaZaki 👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- No linter errors. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️
- Awesome UI. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**